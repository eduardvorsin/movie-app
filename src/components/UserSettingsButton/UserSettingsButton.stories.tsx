import type { Meta, StoryObj } from '@storybook/react';
import UserSettingsButton, { Props } from './UserSettingsButton';
import { useTranslation } from 'react-i18next';

const meta: Meta<typeof UserSettingsButton> = {
	title: 'UI/UserSettingsButton',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the component'
		},
		testId: {
			description: 'Id for testing the component',
		},
		dictionary: {
			control: false,
			description: 'An object with keys for localization of the component',
		},
	},
	component: UserSettingsButton,
	parameters: {
		docs: {
			description: {
				component: 'The UserSettingsButton component is a web interface element that provides user access to their account settings or personal preferences in the application.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof UserSettingsButton>;

const UserSettingsButtonWithHooks = (props: Omit<Props, 'dictionary'>) => {
	const { t } = useTranslation('common');
	const dictionary = {
		button: t('userSettingsButton.button'),
		themeTitle: t('userSettingsButton.themeTitle'),
		languageTitle: t('userSettingsButton.languageTitle'),
		languageSelect: { label: t('languageSelect.label') },
		themeToggle: { label: t('themeToggle.label') },
	};

	return (
		<UserSettingsButton
			{...props}
			dictionary={dictionary}
		/>
	);
}

export const Default: Story = {
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
	render: (args) => (
		<div className='flex justify-center'>
			<UserSettingsButtonWithHooks {...args} />
		</div>
	),
};