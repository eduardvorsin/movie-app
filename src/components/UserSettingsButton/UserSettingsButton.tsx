'use client';

import { GeneralProps } from '@/types/shared';
import Button from '../Button/Button';
import { MouseEventHandler, useId, useState } from 'react';
import Popup from '../Popup/Popup';
import ThemeToggle from '../ThemeToggle/ThemeToggle';
import Title from '../Title/Title';
import LanguageSelect from '../LanguageSelect/LanguageSelect';

export type Props = {
	dictionary: {
		button: string,
		themeTitle: string,
		languageTitle: string,
		languageSelect: Record<'label', string>,
		themeToggle: Record<'label', string>,
	},
} & GeneralProps;

export default function UserSettingsButton({
	className,
	testId,
	dictionary,
}: Props) {
	const [isOpen, setIsOpen] = useState<boolean>(false);
	const id = useId();
	const clickHandler: MouseEventHandler<HTMLButtonElement> = () => {
		setIsOpen((prevState) => !prevState);
	};

	return (
		<Popup
			className='w-[2.25rem] h-[2.25rem]'
			id={id}
			isOpen={isOpen}
			placement='bottom-end'
			offsetX={10}
			offsetY={10}
			lockScroll
			testId={testId}
			trigger={
				<Button
					className={className}
					iconButton
					appearance='secondary'
					onClick={clickHandler}
					aria-haspopup='menu'
					aria-expanded={isOpen}
					aria-controls={id}
					aria-owns={id}
				>
					<svg className='p-2 w-[2.25rem] h-[2.25rem] fill-current' viewBox='0 0 32 32'>
						<use href={'/assets/icons/settings.svg#settings'}></use>
					</svg>
					{dictionary.button}
				</Button>
			}
		>
			<section className='mb-5'>
				<Title
					className='mb-2'
					weight={500}
					as='h2'
					level={6}
				>
					{dictionary.themeTitle}
				</Title>
				<ThemeToggle dictionary={dictionary.themeToggle} />
			</section>

			<section>
				<Title
					className='mb-2'
					weight={500}
					as='h2'
					level={6}
				>
					{dictionary.languageTitle}
				</Title>
				<LanguageSelect dictionary={dictionary.languageSelect} />
			</section>
		</Popup>
	);
};