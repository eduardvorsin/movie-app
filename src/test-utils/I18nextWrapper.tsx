import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;