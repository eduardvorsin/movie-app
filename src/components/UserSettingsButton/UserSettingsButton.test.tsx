import { render, screen } from "@testing-library/react";
import UserSettingsButton from "./UserSettingsButton";
import i18next from "@/i18n/client";
import I18nextWrapper from "@/test-utils/I18nextWrapper";

const dictionary = {
	button: 'Open settings',
	themeTitle: 'The theme of the site',
	languageTitle: 'Language Settings',
	languageSelect: { label: 'Language selection' },
	themeToggle: { label: 'Light and dark theme switch' },
};

describe('UserSettingsButton tests', () => {
	it('is rendered correctly', () => {
		render(
			<UserSettingsButton
				dictionary={dictionary}
				testId='test-user-settings-button'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-user-settings-button')).toBeInTheDocument();
	});

	it('is a basic snapshot', () => {
		render(
			<UserSettingsButton
				dictionary={dictionary}
				testId='test-user-settings-button'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-user-settings-button')).toMatchSnapshot();
	});
});

describe('UserSettingsButton integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('userSettingsButton.button'));

		const dictionary = {
			button: i18next.t('userSettingsButton.button'),
			themeTitle: i18next.t('userSettingsButton.themeTitle'),
			languageTitle: i18next.t('userSettingsButton.languageTitle'),
			languageSelect: { label: i18next.t('languageSelect.label') },
			themeToggle: { label: i18next.t('themeToggle.label') },
		};
		render(
			<UserSettingsButton
				dictionary={dictionary}
				testId='test-user-settings-button'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLButtonElement>(translation)).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('userSettingsButton.button'));

		const dictionary = {
			button: i18next.t('userSettingsButton.button'),
			themeTitle: i18next.t('userSettingsButton.themeTitle'),
			languageTitle: i18next.t('userSettingsButton.languageTitle'),
			languageSelect: { label: i18next.t('languageSelect.label') },
			themeToggle: { label: i18next.t('themeToggle.label') },
		};
		render(
			<UserSettingsButton
				dictionary={dictionary}
				testId='test-user-settings-button'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLButtonElement>(translation)).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');

		const dictionary = {
			button: i18next.t('userSettingsButton.button'),
			themeTitle: i18next.t('userSettingsButton.themeTitle'),
			languageTitle: i18next.t('userSettingsButton.languageTitle'),
			languageSelect: { label: i18next.t('languageSelect.label') },
			themeToggle: { label: i18next.t('themeToggle.label') },
		};
		render(
			<UserSettingsButton
				dictionary={dictionary}
				testId='test-user-settings-button'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLDivElement>('test-user-settings-button')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');

		const dictionary = {
			button: i18next.t('userSettingsButton.button'),
			themeTitle: i18next.t('userSettingsButton.themeTitle'),
			languageTitle: i18next.t('userSettingsButton.languageTitle'),
			languageSelect: { label: i18next.t('languageSelect.label') },
			themeToggle: { label: i18next.t('themeToggle.label') },
		};
		render(
			<UserSettingsButton
				dictionary={dictionary}
				testId='test-user-settings-button'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLDivElement>('test-user-settings-button')).toMatchSnapshot();
	});
});