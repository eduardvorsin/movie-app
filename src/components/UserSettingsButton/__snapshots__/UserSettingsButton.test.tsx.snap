// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSettingsButton integration tests is a snapshot with English localization 1`] = `
<div
  class="relative w-[2.25rem] h-[2.25rem]"
  data-testid="test-user-settings-button"
  id=":r4:"
>
  <button
    aria-controls=":r4:"
    aria-expanded="false"
    aria-haspopup="menu"
    aria-owns=":r4:"
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   "
  >
    <span
      class="inline-flex items-center"
    >
      <svg
        class="p-2 w-[2.25rem] h-[2.25rem] fill-current"
        viewBox="0 0 32 32"
      >
        <use
          href="/assets/icons/settings.svg#settings"
        />
      </svg>
      Open settings
    </span>
  </button>
</div>
`;

exports[`UserSettingsButton integration tests is a snapshot with Russian localization 1`] = `
<div
  class="relative w-[2.25rem] h-[2.25rem]"
  data-testid="test-user-settings-button"
  id=":r5:"
>
  <button
    aria-controls=":r5:"
    aria-expanded="false"
    aria-haspopup="menu"
    aria-owns=":r5:"
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   "
  >
    <span
      class="inline-flex items-center"
    >
      <svg
        class="p-2 w-[2.25rem] h-[2.25rem] fill-current"
        viewBox="0 0 32 32"
      >
        <use
          href="/assets/icons/settings.svg#settings"
        />
      </svg>
      Открыть настройки
    </span>
  </button>
</div>
`;

exports[`UserSettingsButton tests is a basic snapshot 1`] = `
<div
  class="relative w-[2.25rem] h-[2.25rem]"
  data-testid="test-user-settings-button"
  id=":r1:"
>
  <button
    aria-controls=":r1:"
    aria-expanded="false"
    aria-haspopup="menu"
    aria-owns=":r1:"
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   "
  >
    <span
      class="inline-flex items-center"
    >
      <svg
        class="p-2 w-[2.25rem] h-[2.25rem] fill-current"
        viewBox="0 0 32 32"
      >
        <use
          href="/assets/icons/settings.svg#settings"
        />
      </svg>
      Open settings
    </span>
  </button>
</div>
`;
